File: Cpp/Maximal Square.cpp
class Solution {
public:
    /**
     * @param matrix: a matrix of 0 and 1
     * @return: an integer
     */
    int maxSquare(vector<vector<int> > &matrix) {
        // write your code here
        int m = matrix.size();
        if (!m) return 0;
        int n = matrix[0].size();
        vector<vector<int> > size(m, vector<int>(n, 0));
        int maxsize = 0;
        for (int j = 0; j < n; j++) {
            size[0][j] = matrix[0][j];
            maxsize = max(maxsize, size[0][j]);
        }
        for (int i = 1; i < m; i++) {
            size[i][0] = matrix[i][0];
            maxsize = max(maxsize, size[i][0]);
        }
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][j] == 1) {
                    size[i][j] = min(size[i - 1][j - 1], min(size[i - 1][j], size[i][j - 1])) + 1;
                    maxsize = max(maxsize, size[i][j]);
                }
            }
        }
        return maxsize * maxsize;
    }
};

File: Python/Maximal-Square.py
class Solution:
    #param matrix: a matrix of 0 and 1
    #return: an integer
    def maxSquare(self, matrix):
        # write your code here
        res = 0
        lenx = len(matrix)
        leny = len(matrix[0])
        multilist = [[0 for col in range(leny)] for row in range(lenx)]
        for i in range(leny):
            if matrix[0][i] == 1:
                multilist[0][i] = 1
                res = 1
        for i in range(lenx):
            if matrix[i][0] == 1:
                multilist[i][0] = 1
                res = 1
        for i in xrange(1, lenx):
            for j in xrange(1, leny):
                if matrix[i][j] == 1:
                    multilist[i][j] = min(multilist[i - 1][j - 1], min(multilist[i][j - 1], multilist[i - 1][j])) + 1
                    res = max(res, multilist[i][j])
        return res * res